include/mqtt/v5_message.hpp

```diff
@@ -604,70 +606,65 @@ class basic_publish_message {
         }
     }
 
-    template <typename Iterator>
-    basic_publish_message(Iterator b, Iterator e) {
-        if (b >= e) throw remaining_length_error();
-        fixed_header_ = static_cast<std::uint8_t>(*b);
+    basic_publish_message(buffer buf) {
+        if (buf.empty())  throw remaining_length_error();
+        fixed_header_ = static_cast<std::uint8_t>(buf.front());
         auto qos = publish::get_qos(fixed_header_);
-        ++b;
+        buf.remove_prefix(1);
 
-        if (b >= e) throw remaining_length_error();
-        auto len_consumed = remaining_length(b, e);
+        if (buf.empty()) throw remaining_length_error();
+        auto len_consumed = remaining_length(buf.begin(), buf.end());
         remaining_length_ = std::get<0>(len_consumed);
         auto consumed = std::get<1>(len_consumed);
 
-        std::copy(b, b + static_cast<typename std::iterator_traits<Iterator>::difference_type>(consumed), std::back_inserter(remaining_length_buf_));
-        b += static_cast<typename std::iterator_traits<Iterator>::difference_type>(consumed);
+        std::copy(
+            buf.begin(),
+            std::next(buf.begin(), static_cast<buffer::difference_type>(consumed)),
+            std::back_inserter(remaining_length_buf_));
+        buf.remove_prefix(consumed);
 
-        if (b + 2 >= e) throw remaining_length_error();
-        std::copy(b, b + 2, std::back_inserter(topic_name_length_buf_));
-        auto topic_name_length = make_uint16_t(b, b + 2);
-        b += 2;
+        if (buf.size() < 2) throw remaining_length_error();
+        std::copy(buf.begin(), std::next(buf.begin(), 2), std::back_inserter(topic_name_length_buf_));
+        auto topic_name_length = make_uint16_t(buf.begin(), std::next(buf.begin(), 2));
+        buf.remove_prefix(2);
 
-        if (b + topic_name_length >= e) throw remaining_length_error();
-        utf8string_check(string_view(&*b, topic_name_length));
-        topic_name_ = as::buffer(&*b, topic_name_length);
-        b += topic_name_length;
+        if (buf.size() < topic_name_length) throw remaining_length_error();
+        utf8string_check(static_cast<mqtt::string_view>(buf).substr(0, topic_name_length));
```

Comment:
A comment here explaining why you cast to mqtt::string_view would help future contributors.

In this case, the reason is to avoid increasing the reference count of buf when .substr() is called.

Reply:
I agree. I will add a comment.